### ISSUE TITLE
when more than one m/z is inside the ROI the m/z can be way off

### ISSUE THREAD

**Issue by stanstrup** [sneumann/xcms#590]
I encountered an issue that happens when more than one m/z value is inside the ROI for a scan. It seems that what XCMS does is take the average m/z value and sum the intensities of all peaks inside the ROI. This can lead to wildly off m/z values when the ROI contains noise.

My use case is that I am targeting certain peaks (by pre-calculating ROIs) for QC purposes and since we had mass stability issues I need to put a pretty wide m/z range to be sure to actually find the peak. But this cases also some tiny noise peaks to be included and the mass returned is then way off.

I could not find the code that actually does this but I calculated by hand that the above is what happens. What I would suggest is to combine the m/z values by intensity weighted mean instead (like it can be done across the peak).
I think this might also be a more gently solution to the orbitrap shoulder issues than the filter we just implemented in Spectra...

**Comment by jorainer** [comment]
What you describe, is this happening during the integration of the detected peak (where the `"mz"`, `"rt"` and `"into"` are calculated)?

The `mzCenterFun` parameter of `CentWaveParam` should actually allow you to define how the m/z of the peak is calculated. The default is `wMean` which *should* be an intensity-weighted mean. Maybe you can test different options (e.g. `"apex"`, `"mean"`) to see what changes?

**Comment by stanstrup** [comment]
`mzCenterFun` defines how the mass is chosen/calculated _between_ the scans. The issues seems to be that if there are multiple m/z peaks _inside_ the same scan (and inside the ROI) it is the simple mean that is used (for the m/z, the intensities are summed), supposedly before any integration.
It is unclear to me when this happens.

An example:
I have a compound with theoretical mass: 522.3554
I create a ROI in a 200 ppm window, i.e. 522.2505 - 522.4595.

In the _apex_ scan I have peaks at:
```
mz            int
522.3558      3e4
522.4548      2e1
```

What I get in the peak table with `mzCenterFun = "apex"` is mz = 522.4053 (= this is the simple mean of the two peaks in the scan). So 95 ppm off even if the main peak is spot on. It is distorted by the small noise peak.
The weighted mean would have been 522.3559 which is just 1 ppm off.

As I said I cannot figure out where this summing/meaning is happening but I guess it is hardcoded somewhere. And with normal data and reasonably narrow ROIs this rarely matters. A case where it does matter would actually be the orbitrap data we just made the filter for. It is the same issue. Multiple peaks in the same scan within the ROI/ppm parameters. If we used the weighted mean there we might not even need the filter.

**Comment by jorainer** [comment]
OK, thanks for checking that. I will dig a little more into the code.

**Comment by jorainer** [comment]
Can you please in addition also try the followin option - just to see what/if something changes:

```r
options(originalCentWave = FALSE)
```

and then run the CentWave peak detection?

**Comment by jorainer** [comment]
Might be that I need to adapt/update some C functions :(

**Comment by stanstrup** [comment]
> Can you please in addition also try the followin option - just to see what/if something changes:
> 
> ```r
> options(originalCentWave = FALSE)
> ```
> 
> and then run the CentWave peak detection?

I did. Same result.

Yeah I was wondering too if it was hiding in some EIC generation step...

**Comment by jorainer** [comment]
OK - so I'll have to dig into the (undocumented) C-code.

**Comment by jorainer** [comment]
During peak detection, centWave extracts the m/z for the peaks in the ROI [here](https://github.com/sneumann/xcms/blob/master/R/do_findChromPeaks-functions.R#L377-L379) . The peak's m/z is then calculated based on these. 

The `getMZ` function is defined [here](https://github.com/sneumann/xcms/blob/master/src/mzROI.c#L495-L546) and in [this line](https://github.com/sneumann/xcms/blob/master/src/mzROI.c#L536) the function calculates the mean m/z if the ROI contains for one scan (spectrum) more peaks (as I assume is true for your case).

**Comment by stanstrup** [comment]
Yes! that must be it. It is here I suggest an intensity weighted average would be much better.

**Comment by jorainer** [comment]
Or simply report the m/z of the maximum intensity peak? Would be computationally less intense.

**Comment by stanstrup** [comment]
Yes that should work OK as well.

**Comment by jorainer** [comment]
I've implemented an intensity weighted average m/z calculation in the [fix_peak_mz](https://github.com/sneumann/xcms/tree/fix_peak_mz) branch. Can you please check if that works for your data?

**Comment by stanstrup** [comment]
Remembered to push? ;)

**Comment by jorainer** [comment]
Ah, sorry, you're right. I forgot to push. Now it's updated.

**Comment by jorainer** [comment]
Hi @stanstrup - did you have the chance to try it? On my data the m/z of the identified chromatographic peaks slightly change (improve?).

**Comment by jorainer** [comment]
This should now be fixed (version >= 3.17.5). Closing the issue now, feel free to re-open if needed.
